File: Algorithms_Part_2/Ex33LongestRepeatedSubstring.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.Queue;
import edu.princeton.cs.algs4.StdOut;

import java.util.Arrays;

public class Ex33LongestRepeatedSubstring {

    private final int length;
    private final String[] suffixes;

    public Ex33LongestRepeatedSubstring(String s) {
        length = s.length();
        suffixes = new String[length];
        for (int i = 0; i < length; i++)
            suffixes[i] = s.substring(i);

        Arrays.sort(suffixes);
    }

    public int[] findLRS() {
        int index = -1;
        int longest = 0;
        for (int i = 1; i < length; i++) {
            int lcp = lcp(suffixes[i - 1], suffixes[i]);
            if (lcp > longest) {
                longest = lcp;
                index = i - 1;
            }
        }
        if (index == -1) return null;
        StdOut.println("index:" + index + " lcp:" + longest);

        Queue<String> longestStrings = new Queue<>();
        String lString = suffixes[index].substring(0, longest);
        StdOut.println("lString:\"" + lString + "\"");
        for (int i = index; i < length; i++) {
            if (suffixes[i].length() > longest && lString.equals(suffixes[i].substring(0, longest)))
                longestStrings.enqueue(suffixes[i]);
        }


        int[] indices = new int[longestStrings.size()];
        int i = 0;
        for (String s : longestStrings)
            indices[i++] = length - s.length();
        Arrays.sort(indices);
        return indices;
    }

    private int lcp(String a, String b) {
        int n = Math.min(a.length(), b.length());
        for (int i = 0; i < n; i++) {
            if (a.charAt(i) != b.charAt(i)) return i;
        }
        return n;
    }

    public static void main(String[] args) {
        In in = new In("tale.txt");
        String test = in.readAll();
        Ex33LongestRepeatedSubstring lrs = new Ex33LongestRepeatedSubstring(test);

        StringBuilder builder = new StringBuilder();
        builder.append(test);
        int[] indices = lrs.findLRS();
        int occurrences = indices == null ? 0 : indices.length;
        if (occurrences > 0) {
            for (int i = occurrences - 1; i >= 0; i--)
                builder.insert(indices[i], '*');
            StdOut.print("Longest repeated substring was found " +
                                 indices.length + " times in indices:");
            for (int i = 0; i < indices.length; i++)
                StdOut.print(" " + indices[i]);
        }
        else StdOut.println("Longest repeated substring was not found");

        StdOut.println();
        StdOut.println(builder.toString());
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex33LongestRepeatedSubstring.py
class LongestRepeatedSubstring:

    def __init__(self, string):
        self.suffixes = []
        for i in range(len(string)):
            self.suffixes.append(string[i:])
        self.suffixes.sort()

    def find_lrs(self):

        def lcp(s1, s2):
            n = min(len(s1), len(s2))
            for ii in range(n):
                if s1[ii] != s2[ii]:
                    return ii
            return n

        max_lcp = 0
        lcp_index = len(self.suffixes)
        for i in range(len(self.suffixes) - 1):
            lcp_i = lcp(self.suffixes[i], self.suffixes[i + 1])
            if lcp_i > max_lcp:
                max_lcp = lcp_i
                lcp_index = i

        indices = []
        substring = self.suffixes[lcp_index][:max_lcp - 1]
        while lcp_index < len(self.suffixes) and \
                substring == self.suffixes[lcp_index][:max_lcp - 1]:
            indices.append(len(self.suffixes) - len(self.suffixes[lcp_index]))
            lcp_index += 1

        indices.sort()
        return indices


def test_longest_repeated_substring():

    with open('tale.txt') as f:
        text = f.read()

    lrs = LongestRepeatedSubstring(text)
    indices = lrs.find_lrs()

    if len(indices) == 0:
        print('Longest repeated substring was not found')
    else:
        print(f'Longest repeated substring was found {len(indices)} times: {indices}')
        builder = text
        for i in range(len(indices) - 1, -1, -1):
            builder = builder[:indices[i]] + '*' + builder[indices[i]:]
        print(builder)


if __name__ == '__main__':
    test_longest_repeated_substring()
